### DISCUSSION TITLE
Starting with Xcode 15, Apple considers using UserDefaults a privacy problem. Matomo save a UserId in UserDefaults

### DISCUSSION THREAD

**Discussion by GerdC** [matomo-org/matomo-sdk-ios#439] [General]
Hi,

On Twitter, there were discussions 
https://twitter.com/joshavant/status/1684687839739748352 that Apple requires declaring reasons why some APIs are used.

One of the APIs that require declaring a reason is UserDefaults: https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_use_of_required_reason_api

Matomo generates a unique ID for the user and saves it to UserDefaults: https://github.com/matomo-org/matomo-sdk-ios/blob/32c2353d5b57f9e144f199a544c7fe1df9561236/MatomoTracker/Visitor.swift#L19

How are you handling this?
What options are available?

Edit:
My understanding of Apples definition of "Tracking" and "Linking", using analytics on a self owned server is not considered tracking or linking, if not shared with other companies and not linked with data from other companies: https://developer.apple.com/videos/play/wwdc2022/10166

**Comment by brototyp** [comment]
Hey @GerdC, thanks for opening this discussion. This is indeed a very interesting point. I have not thought about this fully, so these thoughts are super fresh and just off the top of my head with the intention to hear thoughts and feedback. These are options I have in mind:

1. Do nothing for now and "wait this out". As so often with Apples rules, the details of them only come to light once they are used by the community.
2. Replace `MatomoUserDefaults` with a `MatomoSettingsStorageType` with a default implementation that still uses `UserDefault`. This would give you all the flexibility to replace this with a persistent storage of your choice while not changing the default behavior right now.
3. Replace `UserDefaults` with another way of storing these settings, e.g. a FileStorage.

Did I forget any?

**Reply by simonnickel** [reply]
As far as I understand, it's just required to describe the used API and reason.

So maybe it's just a setup step to configure the right information in the project.
